### ISSUE TITLE
Implementing a Custom Receiver

### ISSUE THREAD

**Issue by iamdriz** [thibauts/node-castv2-client#17]
I have built a custom receiver that I use with the Google Cast Extension for Chrome.

But can I use this with this package? Your docs only mention the default receiver...

**Comment by joelongstreet** [comment]
I’ve almost successfully implemented a custom receiver, but I’m stuck on getting the receiver to transmit messages back to the sender.

I think I should be able to call something like this on the client side receiver:

```
window.messageBus.broadcast(JSON.stringify({
  type    : 'MEDIA_STATUS',
  status  : { playerState : 'DONE' }
}));
```

But I can’t pick up my sent event using the sample code you’ve provided in the Readme. The below never triggers:

```
player.on('status', function(status) {
  console.log('status broadcast playerState=%s', status.playerState);
});
```

I’ve managed to trace this somewhat and notice it’s getting caught up in the request-response controller because there’s no requestId. If a requestId is included with the broadcast from the client, the request-response controller will die because there’s no callback defined.

Feel like I’m really close. Any idea on what’s wrong here?

Thanks,

Joe

**Comment by thibauts** [comment]
This is a channel problem. You are broadcasting on a channel (I can't tell
what channel from your code), and listening on another one in your node
code. I can't help you as much as I'd like to right now, but you could
check the protocol description in the castv2 repo to learn more about how
channels work. I'll be back at a keyboard in a few days if you don't find a
solution in the meantime!

**Comment by thibauts** [comment]
Have you found a solution ? if not, can you show me the code that creates `window.messageBus` ?

**Comment by joelongstreet** [comment]
Thanks for your help @thibauts 

Unfortunately I wasn't able to resolve my channel issue. I'm a little unclear on the differences between a channel and a message bus namespace. Wouldn't any communication on the namespace would be able be picked up by the node server?

The messageBus on the client is created via:
`window.messageBus = castReceiverManager.getCastMessageBus('urn:x-cast:com.barkley.moonshot.atmosphere');`. If you're interested, I've kept the project open source, and the client can be viewed [here](https://github.com/MoonshotLab/MissionControl/blob/master/public/media-receiver.js).

I took a look at the [castv2-youtube](https://github.com/xat/castv2-youtube) project and used it to create my own custom sender. I didn't change a whole lot, but updated the namespace and the app_id. The receiver controller initialization looks like:

```
function MediaReceiverController(client, sourceId, destinationId) {
  RequestResponseController.call(this, client, sourceId, destinationId, 'urn:x-cast:com.barkley.moonshot.atmosphere');
  this.once('close', onclose);
  var self = this;
  function onclose() {
    self.stop();
  }
}
```

The sender is also open source and I might try to publish it as a little lib if I can get all this working. [The sender controller](https://github.com/MoonshotLab/MissionControl/blob/master/lib/media-receiver-controller.js) and [sender](https://github.com/MoonshotLab/MissionControl/blob/master/lib/media-receiver.js)

**Comment by thibauts** [comment]
Did you try to register a 'message' listener in your controller ?

```javascript
function MediaReceiverController(client, sourceId, destinationId) {
  RequestResponseController.call(this, client, sourceId, destinationId, 'urn:x-cast:com.barkley.moonshot.atmosphere');
  this.once('close', onclose);
  var self = this;
  function onclose() {
    self.stop();
  }
  this.on('message', function(message) {
    console.log(message);
  });
}
```

Each controller abstracts a channel / namespace (they're essentially the same thing). So calls on a controller's methods will result in messages sent on this namespace, and messages broadcast on a namespace will result in a 'message' event *on the controller* handling this namespace. The standard media controller handles its status messages [here](https://github.com/thibauts/node-castv2-client/blob/master/lib/controllers/media.js#L15-L21).

The YouTube sender doesn't listen to broadcasts on its protocol channel `urn:x-cast:com.google.youtube.mdx` which is only used to load media (see [this](https://github.com/xat/castv2-youtube/blob/master/YoutubeController.js)). Instead, it relies on the fact the YouTube app uses the standard media protocol / namespace to manage media playback (see [this](https://github.com/xat/castv2-youtube/blob/master/Youtube.js#L7-L21)). If you look in the methods below the hilighted code you'll see only the `load` method is routed to the youtube namespace.

The `DefaultMediaReceiver` sender provided in this repo also relays the media namespace status events  on the player object [here](https://github.com/thibauts/node-castv2-client/blob/master/lib/senders/default-media-receiver.js#L9-L17).

Basically, sender apps create a bunch of controllers each handling a specific namespace. Events on these namespaces will be fired *on the controller instance* and it's your duty to relay them as needed on the sender app instance.

**Comment by joelongstreet** [comment]
Oooooohhhhh, whoops. This makes much more sense now.

I wasn't listening to messages in the controller, so I wasn't receiving them :).
